kroxylicious-api/src/main/java/io/kroxylicious/proxy/authentication/Subject.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.authentication;
+
+import java.util.Set;
+
+/**
+ * <p>Represents an actor in the system.
+ * Subjects are composed of a possibly-empty set of identifiers represented as {@link Principal} instances.
+ * An anonymous actor is represented by a Subject with an empty set of principals.
+ * As a convenience, {@link Subject#anonymous()} returns such a subject.
+ * </p>
+ *
+ * <p>The principals chosen depend on the calling code but in general might comprise the following:</p>
+ * <ul>
+ * <li>information proven by a client, such as a SASL authorized id,</li>
+ * <li>information known about the client, such as the remote peer's IP address,</li>
+ * <li>information provided by the client, such as its Kafka client id</li>
+ * </ul>
+ * <p>
+ *     <strong>Security best practice says you should only trust information that's proven about the client.</strong>
+ *     However, it is sometimes useful to have access to the other information for making authorization decisions.
+ *     An example would be to deny a misbehaving client application identified by a
+ *     (authorized id, client id)-pair from connecting to a cluster while the underlying problem is fixed.
+ *     Such a decision narrowed access to a client based on untrusted information (the client id)
+ *     to a client which would otherwise be allowed access (based on the SASL authorized id) on
+ *     the premise that such a client is not malicious.
+ *     Crucially, it does not expand access to clients based on that untrusted information.
+ * </p>
```

Comment:
I'm not entirely clear what point your trying to make here. You start with point about best practice then give an example that contradicts that. 

What I think your saying is:
 - Only trust things the client tells you that can be validated externally
 - If information can't be validated default to CLOSED or DENY. i.e. fail closed
 - Only use trusted information when making an ALLOW decision 
 - Un-trusted information can be used to make a DENY decision

That makes me wonder if `Principal` or `Subject` needs some kind of `trusted` / `validated` flag? 

One argument is that a principal should only exist if it has been validated and can be trusted. However that argument makes me wonder how consumers of *this* API are supposed to make use of the advice.